openverse_catalog/dags/providers/provider_api_scripts/inaturalist.py

```diff
@@ -0,0 +1,187 @@
+        self.media_stores["image"].buffer_length = 10_000
+        self.batch_limit = 10_000
+
+        self.sql_template = dedent(
+            (SCRIPT_DIR / "05_export_to_json_template.sql")
+            .read_text()
+            .replace("batch_limit", str(self.batch_limit))
+        )
+
+    def get_next_query_params(self, old_query_params=None, **kwargs):
+        if old_query_params is None:
+            return {"offset_num": 0}
+        else:
+            next_offset = old_query_params["offset_num"] + self.batch_limit
+            return {"offset_num": next_offset}
+
+    def get_response_json(self, query_params: Dict):
+        """
+        Call the SQL to pull json from Postgres, where the raw data has been loaded.
+        """
+        offset_num = str(query_params["offset_num"])
+        sql_string = self.sql_template.replace("offset_num", offset_num)
+        return self.pg.get_records(sql_string)
+
+    def get_batch_data(self, response_json):
+        if response_json:
+            return [dict(item[0]) for item in response_json]
+        return None
+
+    def get_record_data(self, data):
+        license_url = data.get("license_url", None)
+        license_info = get_license_info(license_url=license_url)
+        if license_info == LicenseInfo(None, None, None, None):
+            return None
+
+        foreign_id = data.get("foreign_id")
+        if foreign_id is None:
+            return None
+
+        return {
+            "foreign_identifier": f"{foreign_id}",
+            "foreign_landing_url": data.get("foreign_landing_url"),
+            "title": data.get("title"),
+            "creator": data.get("creator"),
+            "image_url": data.get("image_url"),
+            "width": data.get("width"),
+            "height": data.get("height"),
+            "license_info": license_info,
+            "filetype": data.get("filetype"),
+            "creator_url": data.get("creator_url"),
+            "raw_tags": data.get("tags").split("; "),
+        }
+
+    def get_media_type(self, record):
+        # This provider only supports Images via S3, though they have some audio files
+        # on site and in the API.
+        return "image"
+
+    def endpoint(self):
+        raise NotImplementedError("Normalized TSV files from AWS S3 means no endpoint.")
+
+    @staticmethod
+    def compare_update_dates(
+        last_success: pendulum.DateTime | None, s3_keys: list, aws_conn_id=AWS_CONN_ID
+    ):
+        # if it was never run, assume the data is new
+        if last_success is None:
+            return
+        s3 = S3Hook(aws_conn_id=aws_conn_id)
+        s3_client = s3.get_client_type()
+        for key in s3_keys:
+            try:
+                last_modified = s3_client.head_object(
+                    Bucket="inaturalist-open-data", Key=key
+                )["LastModified"]
+            except Exception as e:
+                logger.error(e)
+                raise AirflowFailException(f"Can't find {key} on s3")
```

Comment:
Right. What are your thoughts on scheduling retries for this task? Max of 3 of them once a week, or max of 28 once a day [I know that's not the same thing], or none and just wait a month and try again?

Reply:
Also, I'm trying with this code: 

```
for key in s3_keys:
            # this will error out if the files don't exist, and bubble up as an
            # informative failure
            last_modified = s3_client.head_object(
                Bucket="inaturalist-open-data", Key=key
            )["LastModified"]
            # if any file has been updated, let's pull them all
            if last_success < last_modified:
                return
        # If no files have been updated, skip the DAG
        raise AirflowSkipException("Nothing new to ingest")
```

And I'm seeing that when there is nothing new to ingest, the dag now gets stuck retrying "drop_loading_table". Should I just put an `IF EXISTS` in [here](https://github.com/WordPress/openverse-catalog/blob/8754cb63c075e2045b91af7bee278566d930286b/openverse_catalog/dags/common/loader/sql.py#L307)? Or would it maybe be better to affirmatively return something (in xcoms? I'm making this up, based on only a small passing familiarity!) to say "yes there is new data, please proceed" and if that doesn't come through, everything else just stops in its tracks?

Reply:
By default, Airflow will retry a task a total of 3 times with 5 minutes in between each run:
https://github.com/WordPress/openverse-catalog/blob/8754cb63c075e2045b91af7bee278566d930286b/openverse_catalog/dags/common/constants.py#L19-L20

I think that should be sufficient! I'm imagining that an _error_ here will be very infrequent/unlikely and will probably require addressing immediately, so it's likely that we'll try to figure out where we went wrong and clear the failing steps once we have. In other words, if this step fails, we'll rerun it again shortly afterwards so I don't think we need to check once a week!

As for your other point, I thought that the generate TSV step would help but you're right, I don't think it resolves the underlying issue 😅 The `IF EXISTS` is such a simple solution, I'll throw that into a separate PR today!

Reply:
Addressed in #634!

Reply:
Thanks so much! I guess what I was thinking was that I don't know that much about how iNaturalist updates their metadata. Is it a human pushing go once a month or is it in airflow? I just looked at the archived metadata though, and check out the dates on the files under `s3://inaturalist-open-data/metadata/`:

```
2021-04-15 18:16:54 4831598102 inaturalist-open-data-20210407.tar.gz
2021-05-20 03:14:10 5295325724 inaturalist-open-data-20210520.tar.gz
2021-07-13 10:58:49 5843976816 inaturalist-open-data-20210713.tar.gz
2021-08-30 20:35:39 6326004812 inaturalist-open-data-20210830.tar.gz
2021-09-21 10:18:24 6520166168 inaturalist-open-data-20210921.tar.gz
2021-10-21 05:53:32 6771547711 inaturalist-open-data-20211021.tar.gz
2021-11-30 14:28:57 6981491162 inaturalist-open-data-20211121.tar.gz
2022-01-27 06:58:19 7312475433 inaturalist-open-data-20220127.tar.gz
2022-03-10 17:00:45 7471017683 inaturalist-open-data-20220227.tar.gz
2022-04-04 13:31:34 7615780100 inaturalist-open-data-20220327.tar.gz
2022-04-27 03:31:45 7905359916 inaturalist-open-data-20220427.tar.gz
2022-06-01 23:05:30 8326761961 inaturalist-open-data-20220527.tar.gz
2022-06-27 04:05:24 8703562223 inaturalist-open-data-20220627.tar.gz
2022-06-27 04:08:14 8703562223 inaturalist-open-data-latest.tar.gz
```

This year it looks like they're actually like clockwork on the 27th of each month. Prior to this year, they were monthly-ish. So, with 3 retries of five minutes, airflow could theoretically have run successfully on 7/15/21, failed on 8/15/21, succeeded on 9/15/21 (picking up the 8/30 data), succeeded on 10/15 (picking up 9/21 data), etc. Unless that's not how the airflow scheduler works? Also, maybe this is really not an issue given the more recent file dates. I'm just trying to think about syncing their monthly schedule with ours.

Reply:
The `@monthly` schedule interval that this provider is set up for means that Airflow will kick it off automatically on UTC 00:00 of the first day of each month. The "Calendar" tab for the DAG gives a good perspective of when each next planned run is - and no human intervention is necessary for kicking it off!

So in the case you've provided, the DAG would have run let's say for the first time on 2021-06-01. The next scheduled run would be 2021-07-01, but that run would skip because there was no newer data. The 2021-08-01 run would be a full run though, since the files were updated on 2021-07-13, etc. So based on this schedule it seems like this DAG will run _nearly_ every month, but there will be some months where it may be skipped.

Reply:
OK, great. Yes, and the first of the month is good relative to the 27th of the prior month when the data is coming out more recently. And so trying 5 and 10 minutes later in case just some weird random thing happens is still worthwhile, because it's more for the weird randomness (some blip at AWS?) than for syncing with iNaturalist's update schedule. Does that sound right?

Reply:
OK, I just committed the changes here, but I am assuming that #634 will get merged first. Fair?